File: mdb/observable_q.lua
local qt = require "qtest"
local O = require "observable"
local list = require "list"

local eq, assert = qt.eq, qt.assert

local Observable, Slot, Func, Log = O.Observable, O.Slot, O.Func, O.Log


--------------------------------
-- Utilities
--------------------------------


local TestSubscriber = Observable:basicNew()

function TestSubscriber:initialize()
   Observable.initialize(self)
   self.invalidations = {}
end

function TestSubscriber:invalidate(ob)
   table.insert(self.invalidations, tostring(ob))
end

function TestSubscriber:assert(obs)
   eq(self.invalidations, list.map(obs, tostring))
end

local sub1 = TestSubscriber:new()
local sub2 = TestSubscriber:new()



--------------------------------
-- Tests
--------------------------------


-- Assertion: Observable:hasInstance() detects child classes.

eq(true, Observable:hasInstance(Observable))
eq(true, Observable:hasInstance(Slot:new(1)))
eq(false, Observable:hasInstance({}))
eq(false, Observable:hasInstance(nil))

eq(true, Slot:hasInstance(Slot:new(1)))
eq(false, Slot:hasInstance(Observable))
eq(false, Slot:hasInstance({}))


-- Assertion: Slots start out "invalid" [they will not propagate invalidations
--     until after a subscriber has called `get`]

local a = Slot:new(23)
eq(a.valid, false)

-- Assertion: Slot does NOT invalidate subscribed listener when already invalid.

a:set(24)
sub1:assert{}

-- Assertion: Slot does NOT invalidate subscribed listeners when not changed.

a:subscribe(sub1)
assert(a:isSubscribed())
a:get()
a:set(24)
sub1:assert{}

-- Assertion: Slot invalidates when valid and changed

a:set(25)
sub1:assert{a}

-- Assertion: multiple subscribers are notified.

a:subscribe(sub2)
a:get()
a:set(26)
sub1:assert{a, a}
sub2:assert{a}


-- Assertion: Slot does NOT invalidate unsubscribed listener, and other
--    subscribed listeners remain subscribed.

a:unsubscribe(sub1)
a:get()
a:set(27)
sub1:assert{a, a}
sub2:assert{a, a}

a:unsubscribe(sub2)
a:get()
a:set(28)
sub2:assert{a, a}



-- Assertion: Func calculates value appropriately, getting values from
--    observable arguments and using other arguments literally.

local a = Slot:new(1)
local b = Slot:new(2)
local sub = TestSubscriber:new()

local calls = 0
local function func(a, b, c)
   calls = calls + 1
   return a * b + c
end

local f = Func:new(func, a, b, 3)

eq(f:get(), 5)
eq(calls, 1)

-- Assertion: Func subscribes to its inputs when it becomes subscribed

assert(not a:isSubscribed())
assert(not b:isSubscribed())
f:subscribe(sub)
assert(f:isSubscribed())
assert(a:isSubscribed())
assert(b:isSubscribed())

-- Assertion: When the Func is valid it does not recalculate

f:get()
calls = 0
eq(5, f:get())
eq(0, calls)

-- Assertion: When a dependency is modified, an update is triggered.

a:set(2)
eq(7, f:get())
eq(1, calls)


-- Assertion: Func unsubscribes from strict AND lazy dependencies when
-- it becomes unsubscribed.

f:unsubscribe(sub)
eq(false, f:isSubscribed())
eq(false, a:isSubscribed())
eq(false, b:isSubscribed())



-- Log

local lsub = TestSubscriber:new()

local log = Log:new()
local v1 = log:get()
eq(v1.len, 0)

log:subscribe(lsub)

log:append("a")
log:append("b")
lsub:assert{log}

local v2 = log:get()
eq(v2.len, 2)
eq(v2.a[1], "a")
eq(v2.a[2], "b")

log:clear()
lsub:assert{log, log}

log:append("c")

local v3 = log:get()
eq(v3.len, 1)
eq(v3.a[1], "c")

-- older versions intact?
eq(v1.len, 0)
eq(v2.len, 2)
eq(v2.a[1], "a")
eq(v2.a[2], "b")

File: jsu/observable_q.js
'use strict';

var O = require('observable.js');
var expect = require('expect.js');
var eq = expect.eq;


//----------------------------------------------------------------
// Mock scheduler
//----------------------------------------------------------------

var callbacks = [];

function sched(cb) {
    callbacks.push(cb);
}

sched.isPending = function () {
    return callbacks.length > 0;
};

sched.call = function () {
    var cb = callbacks.shift();
    cb && cb();
};

sched.flush = function () {
    while (sched.isPending())
        sched.call();
};

var act = O.createActivator(sched);


//----------------------------------------------------------------


// Assertion: Activator.activate(fn, args...) constructs O.Func, subscribes,
//     and calls it.

var ov = O.slot(1);
var count = 0;
var dereg = act.activate(function (a) { count += a; }, ov);
eq(1, count);
eq(false, sched.isPending());

// Assertion: Slot invalidates when valid and changed

ov.setValue(7);
eq(true, sched.isPending());
sched.flush();
eq(8, count);

// Assertion: Slot does NOT invalidate subscribed listeners when not changed

ov.setValue(7);
eq(false, sched.isPending());

// Assertion: Activator.deactivate() unsubscribes to the deactivated object.
// Assertion: Slot does NOT invalidate unsubscribed listener

dereg();
ov.setValue(1);
eq(false, sched.isPending());


// Assertion: Slot does NOT invalidate subscribed listener when already invalid

dereg = act.activate(function (a) { count += a; }, ov);
sched.flush();
eq(true, ov.valid);
ov.valid = false;
ov.setValue(99);
eq(false, sched.isPending());


(function (){
     // Assertion: Func calculates value appropriately, getting values from
     //    observable arguments and using other arguments literally.

     var oa = O.slot(1);
     var f = O.func(function (a, b) { return a+b; }, oa, 2);

     var k = 0;
     var dereg = act.activate(function (v) {
                                   k= v;
                              }, f);
     eq(3, k);

     oa.setValue(3);
     eq(3, k);

     // Assertion: Func subscribes to its inputs when it becomes subscribed

     sched.flush();
     eq(5, k);

     // Assertion: func recognizes constant values

     var f2 = O.func(function (a,b) { return a + b; }, 1, 2);
     eq(3, f2);

}());


// Assertion: Activator.activate(observable) subscribes to the observable.
(function (){
     var ov = O.slot(1);
     var dereg = act.activate(ov);
     eq(1, ov.subs.length);
     dereg();
     eq(0, ov.subs.length);
}());


//----------------------------------------------------------------
// `func` values with lazy and static dependencies
//----------------------------------------------------------------

(function (){

     var oa = O.slot(false);
     var ob = O.slot('b');
     var fValue = 0;
     var fbCount = 0;
     var f = O.func(function (a, fb, bArg) { return (fValue = (a && fb(a))); },
                    oa,
                    function (arg) {
                        ++fbCount;
                        return (typeof arg == 'number') ? arg : ob;
                    });

     // Assertion: Activation computes the initial value.
     var actCount = 0;
     var dereg = act.activate(function () { ++actCount; }, f);
     eq(false, fValue);
     eq(fValue, f.getValue());
     eq(1, actCount);

     // Assertion: When a static dependency is modified, an update is triggered.
     oa.setValue(0);
     sched.flush();
     eq(0, fValue);
     eq(fValue, f.getValue());
     eq(2, actCount);

     // Assertion: When an UNUSED lazy dependency is invalidated, NO update
     // is triggered.
     ob.setValue('OB');
     sched.flush();
     eq(2, actCount);

     // Assertion: When an lazy dependency is accessed and returns a raw value,
     // the raw function receives that value.
     oa.setValue(1);
     sched.flush();
     eq(1, fValue);
     eq(fValue, f.getValue());
     eq(3, actCount);
     eq(1, fbCount);

     // Assertion: When an lazy dependency is accessed and returns an observable,
     // the raw function receives the extracted value.
     oa.setValue('x');
     sched.flush();
     eq('OB', fValue);
     eq(4, actCount);
     eq(2, fbCount);

     // Assertion: The returned observable is subscribed to, and the
     // observable (not its value) is cached.
     var obOld = ob;
     ob = O.slot('OBNew');

     obOld.setValue('OB2');
     sched.flush();
     eq('OB2', fValue);
     eq(2, fbCount);    // no additional calls to fb()


     // Assertion: Calling the thunk with different args discards the cached
     // observable.
     oa.setValue('y');
     sched.flush();
     eq('OBNew', fValue);
     eq(3, fbCount);

     actCount = 0;
     obOld.setValue('-');
     sched.flush();
     eq(0, actCount);

     ob.setValue('xx');
     sched.flush();
     eq('xx', fValue);
     eq(1, ob.subs.length);

     // Assertion: Func unsubscribes from strict AND lazy dependencies when
     // it becomes unsubscribed.

     dereg();
     eq(0, ob.subs.length);
     eq(0, oa.subs.length);

}());
